packages/teleport/src/Discover/Kubernetes/LoginTrait/LoginTrait.tsx

```diff
@@ -0,0 +1,288 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { useState } from 'react';
+import styled from 'styled-components';
+import { Text, Box, Indicator } from 'design';
+import * as Icons from 'design/Icon';
+
+import useTeleport from 'teleport/useTeleport';
+import {
+  SelectCreatable,
+  Option,
+} from 'teleport/Discover/Shared/SelectCreatable';
+import { AccessInfo } from 'teleport/Discover/Shared/AccessInfo';
+
+import {
+  Header,
+  HeaderSubtitle,
+  ActionButtons,
+  ButtonBlueText,
+} from '../../Shared';
+
+import { useLoginTrait, State } from './useLoginTrait';
+
+import type { AgentStepProps } from '../../types';
+
+const resourceName = 'Kubernetes';
+const traitDescription = 'users and groups';
+
+export default function Container(props: AgentStepProps) {
+  const ctx = useTeleport();
+  const state = useLoginTrait({ ctx, props });
+
+  return <LoginTrait {...state} />;
+}
+
+export function LoginTrait({
+  attempt,
+  nextStep,
+  dynamicGroups,
+  staticGroups,
+  dynamicUsers,
+  staticUsers,
+  fetchLoginTraits,
+  canEditUser,
+  isSsoUser,
+}: State) {
+  const [groupInputValue, setGroupInputValue] = useState('');
+  const [groups, setGroups] = useState<Option[]>(() => {
+    const fixedGroups = staticGroups.map(l => ({
+      value: l,
+      label: l,
+      isFixed: true,
+    }));
+    const groups = dynamicGroups.map(l => ({
+      value: l,
+      label: l,
+    }));
+    return [...fixedGroups, ...groups];
+  });
+
+  const [userInputValue, setUserInputValue] = useState('');
+  const [users, setUsers] = useState<Option[]>(() => {
+    const fixedUsers = staticUsers.map(l => ({
+      value: l,
+      label: l,
+      isFixed: true,
+    }));
+    const users = dynamicUsers.map(l => ({
+      value: l,
+      label: l,
+    }));
+    return [...fixedUsers, ...users];
+  });
+
+  const hasUsersOrGroups = groups.length > 0 || users.length > 0;
+  const canAddLoginTraits = !isSsoUser && canEditUser;
+
+  function onProceed() {
+    // Take out duplications and static users and groups.
+    const newDynamicUsers = new Set<string>();
+    users.forEach(o => {
+      if (!staticUsers.includes(o.value)) {
+        newDynamicUsers.add(o.value);
+      }
+    });
+
+    const newDynamicGroups = new Set<string>();
+    users.forEach(o => {
+      if (!staticGroups.includes(o.value)) {
+        newDynamicGroups.add(o.value);
+      }
+    });
+
+    nextStep([...newDynamicUsers], [...newDynamicGroups]);
+  }
+
+  function handleGroupKeyDown(event) {
```

Comment:
Can you add types to handler functions? Typescript will then check of ensure correct usage of SelectCreatable